sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RegionScopedSessionContainer.java

```diff
@@ -0,0 +1,610 @@
+    //      [pkRangeId1 -> {global: <global session token for pkRangeId1>; region1: (GLSN, LLSN, <session token only if first preferred region>); region2...}]
+    private final ConcurrentHashMap<Long, PartitionScopedRegionLevelProgress> collectionResourceIdToPartitionScopedRegionLevelProgress = new ConcurrentHashMap<>();
+
+    // bloom filter which stores an entry which corresponds to a triple of [collectionRid, EPK hash, region]
+    private final PartitionKeyBasedBloomFilter partitionKeyBasedBloomFilter;
+
+    private final ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    private final ReentrantReadWriteLock.ReadLock readLock = readWriteLock.readLock();
+
+    // 1. a write lock is acquired when the collection has not been cached by the session container yet
+    // 2. the write lock is only released when the collection has been recorded along with any session specific
+    // information for a partition in that collection
+    // 3. once a collection has been recorded, thread-safe access is delegated to individual concurrent hashmaps
+    private final ReentrantReadWriteLock.WriteLock writeLock = readWriteLock.writeLock();
+
+    private final ConcurrentHashMap<String, Long> collectionNameToCollectionResourceId = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<Long, String> collectionResourceIdToCollectionName = new ConcurrentHashMap<>();
+    private final String hostName;
+    private boolean disableSessionCapturing;
+    private final GlobalEndpointManager globalEndpointManager;
+    private final AtomicReference<String> firstPreferredReadableRegionCached;
+    private final String regionScopedSessionCapturingOptionsAsString;
+
+    public RegionScopedSessionContainer(final String hostName, boolean disableSessionCapturing, GlobalEndpointManager globalEndpointManager) {
+        this.hostName = hostName;
+        this.disableSessionCapturing = disableSessionCapturing;
+        this.globalEndpointManager = globalEndpointManager;
+        this.firstPreferredReadableRegionCached = new AtomicReference<>(StringUtils.EMPTY);
+        this.partitionKeyBasedBloomFilter = new PartitionKeyBasedBloomFilter();
+        this.regionScopedSessionCapturingOptionsAsString = stringifyConfig();
+    }
+
+    public RegionScopedSessionContainer(final String hostName, boolean disableSessionCapturing) {
+        this(hostName, disableSessionCapturing, null);
+    }
+
+    public RegionScopedSessionContainer(final String hostName) {
+        this(hostName, false, null);
+    }
+
+    public String getHostName() {
+        return this.hostName;
+    }
+
+    @Override
+    public void setDisableSessionCapturing(boolean value) {
+        this.disableSessionCapturing = value;
+    }
+
+    @Override
+    public boolean getDisableSessionCapturing() {
+        return this.disableSessionCapturing;
+    }
+
+    @Override
+    public String getSessionToken(String collectionLink) {
```

Comment:
Hmm... I kept this because there are some fundamental tests being done with this method. Like the following:
- What is the content of the session container for a given collection - can the content be matched if the name-based link or the alt-link are passed? Can this be matched across two different clients? 
- Another test is we have a multi-partition container - so is the document creation / read return a session token which is also a substring of the concatenated session token or not? Can the global session token / concatenated session token be used with request options to read the document etc.

And this was done for `SessionContainer` so thought of extending to `RegionScopedSessionContainer`.